### ISSUE TITLE
DNSSEC Not Working as Expected

### ISSUE THREAD

**Issue by ngocnha** [hickory-dns/hickory-dns#2514]
Upon successful DNSSEC configuration on my hickory-dns server, I anticipated receiving an NSEC record in response to a query for a non-existent RRset. However, the server is only returning an SOA record.
I'm certain DNSSEC is configured correctly. I'm using the latest hickory-dns version pulled from the git repository. What could I be missing?

**Comment by djc** [comment]
Do you have a minimal reproduction?

**Comment by ngocnha** [comment]
I've configured a server on ip addr 45.63.4.93. You can query it to verify.

dig @45.63.4.93 -p 1053 abcdef.pro-service.top A +dnssec +multiline

**Comment by marcus0x62** [comment]
This is related to the issue(s) addressed in #2502.  Right now, the recursor does not do any dnssec processing of any error response (which include negative responses.)

There are also some bugs related to nsec3 record validation that are not addressed in that PR, but for the basic case of properly delegated secure zone returning a nodata response with an nsec proof, 2502 should provide a complete fix.

**Comment by marcus0x62** [comment]
Hi @ngocnha, I've merged #2502 into our main branch. Please check this out and see if it fixes the problem you are seeing.  Note: the domain you included in your example does not appear to return NSEC records, at least not for abcdef.pro-service.top.  If you query against google dns, it will return an EDE error code 12 - NSEC Missing. I rebuilt a server from the current upstream and verified it *did* return NSEC records for another domain which I know is configured correctly.

**Comment by ngocnha** [comment]
I pulled the code from the main branch and built it. I configured DNSSEC according to the example template, and it can automatically generate RRSIG records, but I don't know why it still can't automatically generate NSEC records.

**Comment by marcus0x62** [comment]
Can you post your named.toml file?

**Comment by ngocnha** [comment]
## This is my config file

[[zones]]
zone = "localhost"
zone_type = "Primary"
file = "default/localhost.zone"

[[zones]]
zone = "0.0.127.in-addr.arpa"
zone_type = "Primary"
file = "default/127.0.0.1.zone"

[[zones]]
zone = "0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.ip6.arpa"
zone_type = "Primary"
file = "default/ipv6_1.zone"

[[zones]]
zone = "255.in-addr.arpa"
zone_type = "Primary"
file = "default/255.zone"

[[zones]]
zone = "0.in-addr.arpa"
zone_type = "Primary"
file = "default/0.zone"

[[zones]]
zone = "pro-service.top"
zone_type = "Primary"
enable_dnssec = true
stores = { type = "sqlite", zone_file_path = "zones/pro-service.top.zone", journal_file_path = "zones/pro-service.top_dnssec_update.jrnl", allow_update = true }

[[zones.keys]]
key_path = "dnssec/zsk_rsa_2048.pem"
algorithm = "RSASHA256"
is_zone_signing_key = true
is_zone_update_auth = false

[[zones.keys]]
# Requires --features=ring
key_path = "dnssec/update/ed25519.priv.pk8"
algorithm = "ED25519"
is_zone_signing_key = false
is_zone_update_auth = true

**Comment by divergentdave** [comment]
I think you may need to add `nx_proof_kind = "nsec"` or `nx_proof_kind = { nsec3 = { iterations = 0 } }` below `enable_dnssec = true` to generate those records. (I haven't tested this myself FWIW)

**Comment by marcus0x62** [comment]
I just tested it. @divergentdave is correct.

**Comment by ngocnha** [comment]
Thank @marcus0x62 and @divergentdave for your support. I think I'm missing something, but I couldn't find this option in the document and examples. I just tried searching based on your suggestion and found it somewhere in the testing code.

**Comment by marcus0x62** [comment]
@ngocnha I don't think you missed anything. I think our documentation is a bit behind the code, and for this case we should probably have a default instead of requiring a selection for each zone. Sorry for the confusion.
